js/AppState.js

```diff
@@ -0,0 +1,39 @@
+import * as Derivable from 'derivable';
+import * as Persistence from 'Persistence';
+import * as History from 'History';
+import assign from 'utils/assign';
+
+/**
+ * We store current slide number in location.hash.
+ */
+const slide = History.location.derive(location =>
+  parseInt(location.hash.slice(1), 10) || 1);
+
+/**
+ * We store current mode in persistent atom (syncs with localStorage).
+ */
+const mode = Persistence.atom('__react_shower_mode__', 'list');
+
+/**
+ * App state is a {slide, mode}
+ *
+ * We expose struct derivation to prevent direct modification of the state.
+ */
+export const state = Derivable.struct({mode, slide});
```

Comment:
I really love this: super idiomatic. We define app state as a composition of different pieces.